src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadataProvider.cs

```diff
@@ -167,10 +167,52 @@ protected virtual DefaultMetadataDetailsCache CreateTypeCacheEntry([NotNull] Mod
 
         private class TypeCache : ConcurrentDictionary<ModelMetadataIdentity, DefaultMetadataDetailsCache>
         {
+            public TypeCache()
+                : base(ModelMetadataIdentityComparer.Instance)
+            {
+            }
         }
 
         private class PropertiesCache : ConcurrentDictionary<ModelMetadataIdentity, DefaultMetadataDetailsCache[]>
         {
+            public PropertiesCache()
+                : base(ModelMetadataIdentityComparer.Instance)
+            {
+            }
+        }
+
+        private class ModelMetadataIdentityComparer : IEqualityComparer<ModelMetadataIdentity>
+        {
+            public static readonly ModelMetadataIdentityComparer Instance = new ModelMetadataIdentityComparer();
+
+            public bool Equals(ModelMetadataIdentity x, ModelMetadataIdentity y)
+            {
+                return
+                    x.ContainerType == y.ContainerType &&
+                    x.ModelType == y.ModelType &&
+                    x.Name == y.Name;
+            }
+
+            public int GetHashCode(ModelMetadataIdentity obj)
+            {
+                unchecked
```

Comment:
From MSDN `Expressions that contain non-constant terms are unchecked by default at compile time and run time.`
https://msdn.microsoft.com/en-us/library/a569z7k8.aspx